doc/design/proposal-pre-import.md

```diff
@@ -0,0 +1,94 @@
+# preImport Proposal
+
+## Problem Statement
+
+With the core resolver now synchronous only, any asynchronous resolution work
+is no longer possible as the resolve function must return an immediate
+resolution.
+
+For loaders that asynchronously obtain resolution information, it would be
+useful to have a new hook to enable these use cases.
+
+## preImport Hook
+
+The hook has the signature:
+
+```ts
+export async function preImport(specifier: string, context: {
+  conditions: string[],
+  topLevel: boolean,
+  parentURL: string | undefined
+});
+```
+
+The `preImport` hook allows for tracking and asynchronous setup work for every
+top-level import operation. It has no return value, although it can return
+a promise to to delay the module pipeline operations. No further hooks will
+be called on that module graph until this hook resolves successfully if present.
+
+The `preImport` hook is called for each top-level import operation by the module
+loader, both for the host-called imports (ie for the main entry) and for dynamic
+`import()` imports. These are distinguished by the `dynamic` context.
+
+All `preImport` hooks for all loaders are run asynchronously in parallel, and
+block any further load operations (ie resolve and load) for the module graph
+being imported until they all complete successfully.
+
+Multiple import calls to the same import specifier will re-call the hook
+multiple times. The first error thrown by the `preImport` hooks will be directly
+returned to the specific import operation as the load failure.
+
+## Example
+
+<details>
+<summary>Import Map Generation</summary>
+
+Consider an import map loader which obtains the import map for a module
+asynchronously:
+
+```ts
+import { Generator } from '@jspm/generator';
+
+// stateful host import map for current session
+let importMap = { imports: {}, scopes: {} };
+
+function isUrl (specifier) {
+  try {
+    new URL(specifier);
+    return true;
+  }
+  catch {
+    return false;
+  }
+}
+
+const isBareSpecifier = id => !id.startsWith('./') && !id.startsWith('../') &&
+    !id.startsWith('/') && !isUrl(id);
+
+export async function preImport(specifier, { conditions, parentURL }) {
+  if (!isBareSpecifier(specifier)) return;
+  const generator = new Generator({
+    // passing the original map extends it
+    inputMap: importMap,
+    baseUrl: parentURL,
+    defaultProvider: 'nodemodules',
+    env: conditions,
+  });
+  await generator.install(specifier);
+  // the new map will then have the new mappings and the old mappings
+  importMap = generator.getMap();
+}
```

Comment:
Since this function doesn’t return anything, how are its results preserved for use in `resolve` later? As part of the internal state of `generator` or something it references?

If that’s the case, it might be good for the example to illustrate this. Maybe a variable could be declared at the top level of the loader that gets referenced by both hooks, and the contents of this variable are how the information determined by `preImport` gets shared with `resolve`.

Reply:
The `importMap` mutable variable is the shared state.

Reply:
Ah, sorry, I missed this. I don’t know if it’s just me or if it’s likely to be nonobvious to others; if the latter, maybe add a comment inside `resolve` to the effect of ``// Reference same `importMap` variable modified within `preImport` ``. That would make the expected usage clear, so people know how this hook is expected to work.